### PR TITLE
Move rank-eval template compilation down to TransportRankEvalAction

### PR DESCRIPTION
**Pull request by MaineC**
Relates to #21777
    
Search templates for rank_eval endpoint so far only worked when sent through
REST end point
    
However we also allow templates to be set through a Java API call to
"setTemplate" on that same spec. This doesn't go through template execution so
fails further down the line.
    
To make this work, moved template execution further down, probably to
TransportRankEvalAction.

@cbuescher would be nice if you could have a look

### CODE REVIEW

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java [hunk 1]
```diff
@@ -73,6 +91,21 @@ protected void doExecute(RankEvalRequest request, ActionListener<RankEvalRespons
             final RankEvalActionListener searchListener = new RankEvalActionListener(listener, qualityTask.getMetric(), querySpecification,
                     partialResults, errors, responseCounter);
             SearchSourceBuilder specRequest = querySpecification.getTestRequest();
+            if (specRequest == null) {
+                Map<String, Object> params = querySpecification.getParams();
+                Script scriptWithParams = new Script(qualityTask.getTemplate().getType(), qualityTask.getTemplate().getLang(),
+                        qualityTask.getTemplate().getIdOrCode(), params);
+                String resolvedRequest = ((BytesReference) (scriptService.executable(scriptWithParams, ScriptContext.Standard.SEARCH)
+                        .run())).utf8ToString();
+                logger.error("RANTRANTRANT" + resolvedRequest);
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java]: left over?

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedRequest.java [hunk 2]
```diff
@@ -101,7 +103,13 @@ public RatedRequest(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeString(specId);
-        testRequest.writeTo(out);
+        if (testRequest != null) {
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedRequest.java]: nit: you could use writeOptionalWriteable() here

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedRequest.java [hunk 3]
```diff
@@ -74,7 +74,9 @@ public RatedRequest(String specId, SearchSourceBuilder testRequest, List<String>
 
     public RatedRequest(StreamInput in) throws IOException {
         this.specId = in.readString();
-        testRequest = new SearchSourceBuilder(in);
+        if (in.readBoolean()) {
+            testRequest = new SearchSourceBuilder(in);
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedRequest.java]: nit: possibly use readOptionalWriteable here

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java [hunk 4]
```diff
@@ -73,6 +91,21 @@ protected void doExecute(RankEvalRequest request, ActionListener<RankEvalRespons
...
+                try (XContentParser subParser = XContentFactory.xContent(resolvedRequest).createParser(resolvedRequest)) {
+                    QueryParseContext parseContext = new QueryParseContext(searchRequestParsers.queryParsers, subParser, parseFieldMatcher);
+                    specRequest = SearchSourceBuilder.fromXContent(parseContext, searchRequestParsers.aggParsers,
+                            searchRequestParsers.suggesters, searchRequestParsers.searchExtParsers);
+                } catch (IOException e) {
+                    listener.onFailure(e);
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java]: Just a question: this means we are failing the whole rank_eval request here, no? I think this is okay, as long as this is what you intended.

**Review by cbuescher — APPROVED** [review]
I left some minor remarks and a question, but I would also be okay with pushing this as is.

**Comment by MaineC** [comment]
@cbuescher updated

**Comment by cbuescher** [comment]
LGTM

**Pull request merged**